teps/0008-api-spec.md

```diff
@@ -0,0 +1,115 @@
+
+The most valuable product of the Tekton project is its API. The Kubernetes implementation is definitely _useful_, as it allows real-world use cases to take advantage of the API, and -- because it's based on Kubernetes -- have a working implementation of it that can run on a variety of hardware setups and cloud providers. But the API is the real product.
+
+The API today is large, sprawling, and relatively fast-moving. This is a _good thing_, as it provides users a plethora of options to describe and run their workloads. But it's also a challenge for new users, and for potential platform providers and implementers. Inclusion of Kubernetes built-in types (`Container`, `PodSpec`, `Volume`, etc.) accounts for the majority of the API's total surface area, even though many of those options are rarely used by users, and may not even be supported on most platforms. Some fields are ignored outright by the Kubernetes-based reference implementation (`imagePullPolicy`), and some are unusable in practice (`livenessProbe`, `terminationMessagePath`). Support for these fields is difficult to reason about in the abstract, particularly in cases where a future implementation might not be running on Kubernetes at all.
+
+By explicitly stating which portions of the API _must_ be supported to be considered a "conformant" Tekton implementation, which are optional nice-to-haves, and which are explicitly not recommended or expected to be supported, the API is easier to understand and adopt for users, operators, platform providers, and the community as a whole.
+
+### Goals
+
+Produce and maintain a document in the [tektoncd/pipeline](https://github.com/tektoncd/pipeline) repo documenting the v1beta1 Tekton Pipelines API, including which fields are required, recommended and not recommended to be supported by a conformant Tekton Pipelines implementation.
+
+This will serve as documentation to end users and client authors (including Tekton's own CLI and Dashboard) about which features it should expect to be supported.
+
+Document the lifecycle of a Tekton Pipelines API feature, through alpha to beta to GA, and possibly separately to being required to satisfy API spec conformance. This process should cover how to communicate to clients and existing implementations that a new field will newly be considered required.
+
+Produce and maintain a conformance test suite which can be run against a candidate Tekton implementation to determine whether it satisfies the documented spec. This test suite should be regularly run against a platform to guard against regressions.
+
+### Non-Goals
+
+Document the [runtime contract](https://tekton.dev/docs/pipelines/container-contract/), or prescribe specific implementations of supporting services such as access control, observability and resource management.
```

Comment:
i would think that conformance tests would be testing some parts of the runtime contract as well - i assume one of the goals of conformance (maybe we could add above if you agree) is to ensure portability b/w implementations (or maybe not?)

or maybe this is something we'd handle in a TEP specific to conformance and you just want to focus on the API spec here?

Reply:
o yeah you mentioned portability in the user stories below, so maybe it makes sense to include

Reply:
I'd like this to be the last TEP for now related to API specification -- the next ones will come as we expand its coverage to include Tasks/Pipelines/PipelineRuns.

This TEP is tldr "we should document and test conformance", if there's something you think we should cover that's not roughly outlined in this TEP, we should push on that in this TEP I think.

Testing the runtime contract is going to be quite a bit more work, I think, and I think it's unlikely that spec-covered features ("run this image with these args") will have divergent behavior between two implementations. If we find that's not the case, then we can document the runtime contract better, and even add it to conformance tests. But I'd like to figure out if it's a problem before we commit to testing the entire runtime contract right out the gate.

wdyt?